### PR TITLE
Better queue expiration

### PR DESCRIPTION
**Pull request by marshalium**
There are two fixes here:

1. Only expire queues if it's been max age since a modifying operation has occurred (remove, add, confirm, or abort) instead of if it's been max age since creation. This will hopefully reduce queue churn for us significantly. Our queues often empty because our workers keep up with them and Kestrel blocking to read from disk every expiration interval to re-read all the queues is bad. It also causes us to drop queue items because if we're still writing while the queue is being expired those items get lost.

2. Prevent expiring the queue when there are still open transactions. Until they've been confirmed open transactions are just as important as any other queue item!

### CODE REVIEW

File: src/main/scala/net/lag/kestrel/PersistentQueue.scala [hunk 1]
```diff
@@ -47,7 +47,10 @@ class PersistentQueue(val name: String, persistencePath: String, @volatile var c
   private var _currentAge: Time = Time.epoch
 
   // time the queue was created
-  private var _createTime = Time.now
+  private val _createTime = Time.now
```
Comment (tehsven) [inline src/main/scala/net/lag/kestrel/PersistentQueue.scala]: why the change to `val`? `_createTime` isn't being mutated anywhere?

Comment (tehsven) [inline src/main/scala/net/lag/kestrel/PersistentQueue.scala]: nvm, I had them backwards

File: src/main/scala/net/lag/kestrel/PersistentQueue.scala [hunk 2]
```diff
@@ -182,15 +187,10 @@ class PersistentQueue(val name: String, persistencePath: String, @volatile var c
...
    */
-  def isReadyForExpiration: Boolean = {
-    // Don't even bother if the maxQueueAge is None
-    if (config.maxQueueAge.isDefined && queue.isEmpty && Time.now > _createTime + config.maxQueueAge.get) {
-      true
-    } else {
-      false
-    }
+   def isReadyForExpiration: Boolean = synchronized {
+    config.maxQueueAge.isDefined && queue.isEmpty && openTransactions.isEmpty && (Time.now > _lastModificationTime + config.maxQueueAge.get)
```
Comment (tehsven) [inline src/main/scala/net/lag/kestrel/PersistentQueue.scala]: `_lastModificationTime` is in ms but `maxQueueAge` is defined in seconds (at least according to the spec) - is there some type conversion going on here?

Comment (tehsven) [inline src/main/scala/net/lag/kestrel/PersistentQueue.scala]: https://github.com/twitter/util
```
val duration1 = 1.second
val duration2 = 2.minutes
duration1.inMillis // => 1000L
```

File: src/main/scala/net/lag/kestrel/PersistentQueue.scala [hunk 3]
```diff
@@ -182,15 +187,10 @@ class PersistentQueue(val name: String, persistencePath: String, @volatile var c
 
   /**
    * Check if this Queue is eligible for expiration by way of it being empty
-   * and its age being greater than or equal to maxQueueAge
+   * and the last queue modification occuring longer than maxQueueAge ago.
```
Comment (tehsven) [inline src/main/scala/net/lag/kestrel/PersistentQueue.scala]: and there are no open transactions

File: src/main/scala/net/lag/kestrel/PersistentQueue.scala [hunk 4]
```diff
@@ -302,6 +302,8 @@ class PersistentQueue(val name: String, persistencePath: String, @volatile var c
    */
   def add(value: Array[Byte], expiry: Option[Time], xid: Option[Int], addTime: Time): Boolean = {
     val future = synchronized {
+      _lastModificationTime = Time.now
+
       if (closed || value.size > config.maxItemSize.inBytes) return false
       if (config.fanoutOnly && !isFanout) return true
       while (queueLength >= config.maxItems || queueSize >= config.maxSize.inBytes) {
```
Comment (tehsven) [inline src/main/scala/net/lag/kestrel/PersistentQueue.scala]: seems like `_lastModificationTime = Time.now` should go after these two early return checks since nothing is mutated if it early returns

File: src/main/scala/net/lag/kestrel/PersistentQueue.scala [hunk 5]
```diff
@@ -365,6 +367,8 @@ class PersistentQueue(val name: String, persistencePath: String, @volatile var c
    */
   def remove(transaction: Boolean): Option[QItem] = {
     val removedItem = synchronized {
+      _lastModificationTime = Time.now
```
Comment (tehsven) [inline src/main/scala/net/lag/kestrel/PersistentQueue.scala]: seems like this should only go in the `else` block

File: src/main/scala/net/lag/kestrel/PersistentQueue.scala [hunk 6]
```diff
@@ -474,6 +480,8 @@ class PersistentQueue(val name: String, persistencePath: String, @volatile var c
 
   def confirmRemove(xid: Int) {
     synchronized {
+      _lastModificationTime = Time.now
```
Comment (tehsven) [inline src/main/scala/net/lag/kestrel/PersistentQueue.scala]: why update here? it's not touching the queue, just the journal and open transactions.

Comment (tehsven) [inline src/main/scala/net/lag/kestrel/PersistentQueue.scala]: as discussed, seems like changing to `_lastAccessTime` and adding appropriate comment of intention, and adding to `peek` would more clearly get at the intended behavior

File: src/test/scala/net/lag/kestrel/PersistentQueueSpec.scala [hunk 7]
```diff
@@ -1089,8 +1089,97 @@ class PersistentQueueSpec extends SpecificationWithJUnit
...
+          val config = new QueueBuilder {
+            keepJournal = false
+            maxQueueAge = 90.seconds
+          }.apply()
+          val q = new PersistentQueue("wu_tang", folderName, config, timer, scheduler)
+          q.setup()
+
+          q.isReadyForExpiration mustEqual false
+
+          q.add("method man".getBytes, None) mustEqual true
```
Comment (tehsven) [inline src/test/scala/net/lag/kestrel/PersistentQueueSpec.scala]: favorite line of this PR

**Comment by tehsven** [comment]
A few comments, but overall looks good

File: src/main/scala/net/lag/kestrel/PersistentQueue.scala (hunk 2 above)
Comment (marshalium) [inline src/main/scala/net/lag/kestrel/PersistentQueue.scala]: `Time.now` and `_lastModificationTime` are `Time`s and `maxQueueAge.get` is a `Duration`. `Time `+ `Duration` is a `Time`.

File: src/main/scala/net/lag/kestrel/PersistentQueue.scala (hunk 3 above)
Comment (marshalium) [inline src/main/scala/net/lag/kestrel/PersistentQueue.scala]: good catch

File: src/main/scala/net/lag/kestrel/PersistentQueue.scala (hunk 5 above)
Comment (marshalium) [inline src/main/scala/net/lag/kestrel/PersistentQueue.scala]: As discussed this needs to go here so that even if there are zero queue items it gets updated.

**Comment by tehsven** [comment]
+1 this is much more clear. looks good!
